src/Build/Graph/GraphBuilder.cs

```diff
@@ -162,27 +160,20 @@ private void AddEdgesFromProjectReferenceItems(Dictionary<ConfigurationMetadata,
 
             HashSet<ProjectGraphNode> GetTransitiveProjectReferencesExcludingSelf(ParsedProject parsedProject)
             {
-                if (transitiveReferenceCache.TryGetValue(parsedProject.GraphNode, out HashSet<ProjectGraphNode> cachedTransitiveReferences))
-                {
-                    return cachedTransitiveReferences;
-                }
-                else
-                {
-                    var transitiveReferences = new HashSet<ProjectGraphNode>();
+                HashSet<ProjectGraphNode> references = new();
+                GetTransitiveProjectReferencesExcludingSelfHelper(parsedProject, references);
+                return references;
+            }
 
-                    foreach (var referenceInfo in parsedProject.ReferenceInfos)
+            void GetTransitiveProjectReferencesExcludingSelfHelper(ParsedProject parsedProject, HashSet<ProjectGraphNode> transitiveReferences)
+            {
+                foreach (ProjectInterpretation.ReferenceInfo referenceInfo in parsedProject.ReferenceInfos)
+                {
+                    if (!transitiveReferences.Contains(allParsedProjects[referenceInfo.ReferenceConfiguration].GraphNode))
                     {
                         transitiveReferences.Add(allParsedProjects[referenceInfo.ReferenceConfiguration].GraphNode);
```

Comment:
You can combine both add and contains into one call, by checking the return value on Add. Avoids a hash ode computation and lookupm